src/commonMain/kotlin/dev/snipme/highlights/Highlights.kt

```diff
@@ -60,22 +69,37 @@ class Highlights private constructor(
     }
 
     fun getHighlights(): List<CodeHighlight> {
-        val highlights = mutableListOf<CodeHighlight>()
         val structure = getCodeStructure()
-        with(structure) {
-            marks.forEach { highlights.add(ColorHighlight(it, theme.mark)) }
-            punctuations.forEach { highlights.add(ColorHighlight(it, theme.punctuation)) }
-            keywords.forEach { highlights.add(ColorHighlight(it, theme.keyword)) }
-            strings.forEach { highlights.add(ColorHighlight(it, theme.string)) }
-            literals.forEach { highlights.add(ColorHighlight(it, theme.literal)) }
-            annotations.forEach { highlights.add(ColorHighlight(it, theme.metadata)) }
-            comments.forEach { highlights.add(ColorHighlight(it, theme.comment)) }
-            multilineComments.forEach { highlights.add(ColorHighlight(it, theme.multilineComment)) }
-        }
-
-        emphasisLocations.forEach { highlights.add(BoldHighlight(it)) }
+        return constructHighlights(structure)
+    }
 
-        return highlights
+    fun getHighlightsAsync(listener: HighlightsResultListener) {
+        try {
+            val errorHandler = CoroutineExceptionHandler { _, exception ->
+                listener.onError(exception)
+            }
+
+            fun runJob() {
+                CoroutineScope(Dispatchers.Default + errorHandler).launch {
```

Comment:
This will leak the job if the process never finishes. I recommend creating a separate API like `codeHighlightState` and passing the coroutine scope within it for processing, where you would emit results in a flow. In general, I am not happy with the java style listener API and not sure why you decided to go this way.

Reply:
I have in mind the compatibility with Dart and Swift and wanted provide the most tech-agnostic API as possible but after testing integrations with Coroutines Multiplatform maybe interface will change